# **Intro to Pandas**

- The Basics
- Exploring a dataset
- Working with nulls
- Transforming the data
- String Methods
- Joins
- Pivot Tables
- Crosstab 
- Group By

Import the pandas module as `pd`

In [1]:
import pandas as pd

In [ ]:
#print(dir(pd))

# **Series**

#### **Create a Series**

- One dimensional data structure
- Ordered, indexed, mutable
- Covenient Aggregations
- has a mutable index, similar to dictionary
- Element by element calculations, without loops

#### **Creating a Series From a List**

In [2]:
lst = [2, 4, 6, 8]
s = pd.Series(lst)

print(lst)
print(s)

[2, 4, 6, 8]
0    2
1    4
2    6
3    8
dtype: int64


In [3]:
lst[3]

8

#### **Creating a Series from a dictionary**

In [4]:
dct = {'A':2,'B':4,'C':6,'D':8,'E':10,'F':12}

a = pd.Series(dct)
print(a)

A     2
B     4
C     6
D     8
E    10
F    12
dtype: int64


In [5]:
a[['C', 'D']]

C    6
D    8
dtype: int64

#### **Convenient Aggregations**

In [6]:
print(s.min())
print(s.max())
print(s.mean())
print(s.median())

2
8
5.0
5.0


#### **Element by element calculations wiuthout loops**

In [7]:
lst = lst * 2
print(lst)

[2, 4, 6, 8, 2, 4, 6, 8]


In [8]:
lst = [2, 4, 6, 8]

In [9]:
lst_times_two = []
s = pd.Series(lst)

for i in lst:
    lst_times_two.append(i * 2)
print(lst_times_two)

[4, 8, 12, 16]


In [10]:
s = pd.Series(lst)
s = (s * 2)
print(s)

0     4
1     8
2    12
3    16
dtype: int64


#### **Changing the index**

In [11]:
s1 = pd.Series(lst, index = ['z','x','e','p'])
print(s1)

z    2
x    4
e    6
p    8
dtype: int64


In [12]:
s1['p']

8

# **Dataframes**

Pandas allows you to create datasets within Python. These datasets can be recalled and used once they are defined. You can do this by including the following parameters:

1. import pandas (you will need to do this each time you close and reopen Python)  
2. name the dataset. 
3. include the variables and their values. 
4. create the dataset as a DataFrame. 
5. print the DataFrame.


In [13]:
empty_df = pd.DataFrame()
print(empty_df)

Empty DataFrame
Columns: []
Index: []


In [14]:
employeedataset = {
    'name': ['John Smith', 'Sally Jones', 'Sam Harrison'],
    'department': ['HR','IT','IR']
    }

myvar = pd.DataFrame(employeedataset)

myvar

name department
0    John Smith         HR
1   Sally Jones         IT
2  Sam Harrison         IR

In [15]:
data = {
    'calories': [520, 480, 490],
    'duration': [50, 40, 45]
    }

myvar = pd.DataFrame(data)

myvar

calories  duration
0       520        50
1       480        40
2       490        45

# **Exploring a dataset**

Import the .csv file with the `pd.read_csv()`

In [16]:
df = pd.read_csv('data/module7_countries.csv')

[For more information pd.read_csv() click here](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)

In [17]:
df

Country                               Region  Population  \
0       Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1           Albania   EASTERN EUROPE                          3581655   
2           Algeria   NORTHERN AFRICA                        32930091   
3    American Samoa   OCEANIA                                   57794   
4           Andorra   WESTERN EUROPE                            71201   
..               ...                                  ...         ...   
230       West Bank   NEAR EAST                               2460492   
231  Western Sahara   NORTHERN AFRICA                          273008   
232           Yemen   NEAR EAST                              21456188   
233          Zambia   SUB-SAHARAN AFRICA                     11502010   
234        Zimbabwe   SUB-SAHARAN AFRICA                     12236805   

     Area (sq. mi.) Pop. Density (per sq. mi.) Coastline (coast/area ratio)  \
0            647500                       48,0                         0,00   
1             28748                      124,6                         1,26   
2           2381740                       13,8                         0,04   
3               199                      290,4                        58,29   
4               468                      152,1                         0,00   
..              ...                        ...                          ...   
230            5860                      419,9                         0,00   
231          266000                        1,0                         0,42   
232          527970                       40,6                         0,36   
233          752614                       15,3                         0,00   
234          390580                       31,3                         0,00   

    Net migration Infant mortality (per 1000 births)  GDP ($ per capita)  \
0           23,06                             163,07               700.0   
1           -4,93                              21,52              4500.0   
2           -0,39                                 31              6000.0   
3          -20,71                               9,27              8000.0   
4             6,6                               4,05             19000.0   
..            ...                                ...                 ...   
230          2,98                              19,62               800.0   
231           NaN                                NaN                 NaN   
232             0                               61,5               800.0   
233             0                              88,29               800.0   
234             0                              67,69              1900.0   

    Literacy (%) Phones (per 1000) Arable (%) Crops (%) Other (%) Climate  \
0           36,0               3,2      12,13      0,22     87,65       1   
1           86,5              71,2      21,09      4,42     74,49       3   
2           70,0              78,1       3,22      0,25     96,53       1   
3           97,0             259,5         10        15        75       2   
4          100,0             497,2       2,22         0     97,78       3   
..           ...               ...        ...       ...       ...     ...   
230          NaN             145,2       16,9     18,97     64,13       3   
231          NaN               NaN       0,02         0     99,98       1   
232         50,2              37,2       2,78      0,24     96,98       1   
233         80,6               8,2       7,08      0,03      92,9       2   
234         90,7              26,8       8,32      0,34     91,34       2   

    Birthrate Deathrate Agriculture Industry Service  
0        46,6     20,34        0,38     0,24    0,38  
1       15,11      5,22       0,232    0,188   0,579  
2       17,14      4,61       0,101      0,6   0,298  
3       22,46      3,27         NaN      NaN     NaN  
4        8,71      6,25         NaN      NaN     NaN  
..        ...      

#### **Display Max and Max Columns**

In [18]:
pd.set_option('display.max_rows', 10)
pd.set_option('display.max_columns', 10)

Import the provided data

In [110]:
df.head(5)

Country                               Region  Population  \
0     Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1         Albania   EASTERN EUROPE                          3581655   
2         Algeria   NORTHERN AFRICA                        32930091   
3  American Samoa   OCEANIA                                   57794   
4         Andorra   WESTERN EUROPE                            71201   

   Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
0          647500                       48,0  ...      46,6     20,34   
1           28748                      124,6  ...     15,11      5,22   
2         2381740                       13,8  ...     17,14      4,61   
3             199                      290,4  ...     22,46      3,27   
4             468                      152,1  ...      8,71      6,25   

  Agriculture  Industry Service  
0        0,38      0,24    0,38  
1       0,232     0,188   0,579  
2       0,101       0,6   0,298  
3         NaN       NaN     NaN  
4         NaN       NaN     NaN  

[5 rows x 20 columns]

[For more information about df.head() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html)

In [20]:
df.tail(1)

Country                               Region  Population  \
234  Zimbabwe   SUB-SAHARAN AFRICA                     12236805   

     Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
234          390580                       31,3  ...     28,01     21,84   

    Agriculture  Industry Service  
234       0,179     0,243   0,579  

[1 rows x 20 columns]

[For more information about df.tail() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.tail.html#pandas.DataFrame.tail)

In [21]:
df.describe() #Only useful for columns recognized as numeric

Population  Area (sq. mi.)  GDP ($ per capita)
count  2.350000e+02    2.350000e+02          234.000000
mean   2.872891e+07    6.165375e+05         9499.145299
std    1.164019e+08    1.836074e+06         9918.790385
min    7.026000e+03    2.000000e+00          500.000000
25%    4.105965e+05    3.488500e+03         1900.000000
50%    4.786994e+06    8.660000e+04         5400.000000
75%    1.749777e+07    4.219110e+05        14850.000000
max    1.313974e+09    1.707520e+07        55100.000000

[For more information about df.describe() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html)

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 20 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Country                             235 non-null    object 
 1   Region                              235 non-null    object 
 2   Population                          235 non-null    int64  
 3   Area (sq. mi.)                      235 non-null    int64  
 4   Pop. Density (per sq. mi.)          235 non-null    object 
 5   Coastline (coast/area ratio)        235 non-null    object 
 6   Net migration                       232 non-null    object 
 7   Infant mortality (per 1000 births)  232 non-null    object 
 8   GDP ($ per capita)                  234 non-null    float64
 9   Literacy (%)                        216 non-null    object 
 10  Phones (per 1000)                   230 non-null    object 
 11  Arable (%)                          232 non-n

[For more information about df.info() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html)

In [23]:
df.columns

Index(['Country', 'Region', 'Population', 'Area (sq. mi.)',
       'Pop. Density (per sq. mi.)', 'Coastline (coast/area ratio)',
       'Net migration', 'Infant mortality (per 1000 births)',
       'GDP ($ per capita)', 'Literacy (%)', 'Phones (per 1000)', 'Arable (%)',
       'Crops (%)', 'Other (%)', 'Climate', 'Birthrate', 'Deathrate',
       'Agriculture', 'Industry', 'Service'],
      dtype='object')

#### **Locate Rows**

In [24]:
df.loc[0:10:5] #this one is special!

Country                               Region  Population  \
0   Afghanistan         ASIA (EX. NEAR EAST)             31056997   
5        Angola   SUB-SAHARAN AFRICA                     12127071   
10      Armenia                  C.W. OF IND. STATES      2976372   

    Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
0           647500                       48,0  ...      46,6     20,34   
5          1246700                        9,7  ...     45,11      24,2   
10           29800                       99,9  ...     12,07      8,23   

   Agriculture  Industry Service  
0         0,38      0,24    0,38  
5        0,096     0,658   0,246  
10       0,239     0,343   0,418  

[3 rows x 20 columns]

#### df.loc Syntax
##### `df.loc[rows, columns]`

##### Display all values in one column

In [25]:
df.loc[:,'Population']

0      31056997
1       3581655
2      32930091
3         57794
4         71201
         ...   
230     2460492
231      273008
232    21456188
233    11502010
234    12236805
Name: Population, Length: 235, dtype: int64

##### **Display a range of columns**

Show Region and Net Migration

In [26]:
df.loc[0:10,['Region','Net migration']]

Region Net migration
0         ASIA (EX. NEAR EAST)                  23,06
1   EASTERN EUROPE                              -4,93
2   NORTHERN AFRICA                             -0,39
3   OCEANIA                                    -20,71
4   WESTERN EUROPE                                6,6
..                                  ...           ...
6               LATIN AMER. & CARIB             10,76
7               LATIN AMER. & CARIB             -6,15
8               LATIN AMER. & CARIB             -6,15
9               LATIN AMER. & CARIB              0,61
10                 C.W. OF IND. STATES          -6,47

[11 rows x 2 columns]

Show Region through net Migration

In [27]:
df.loc[0:10,'Region':'Net migration']

Region  Population  Area (sq. mi.)  \
0         ASIA (EX. NEAR EAST)             31056997          647500   
1   EASTERN EUROPE                          3581655           28748   
2   NORTHERN AFRICA                        32930091         2381740   
3   OCEANIA                                   57794             199   
4   WESTERN EUROPE                            71201             468   
..                                  ...         ...             ...   
6               LATIN AMER. & CARIB           13477             102   
7               LATIN AMER. & CARIB           69108             443   
8               LATIN AMER. & CARIB           69108             443   
9               LATIN AMER. & CARIB        39921833         2766890   
10                 C.W. OF IND. STATES      2976372           29800   

   Pop. Density (per sq. mi.) Coastline (coast/area ratio) Net migration  
0                        48,0                         0,00         23,06  
1                       124,6                         1,26         -4,93  
2                        13,8                         0,04         -0,39  
3                       290,4                        58,29        -20,71  
4                       152,1                         0,00           6,6  
..                        ...                          ...           ...  
6                       132,1                        59,80         10,76  
7                       156,0                        34,54         -6,15  
8                       156,0                        34,54         -6,15  
9                        14,4                         0,18          0,61  
10                       99,9                         0,00         -6,47  

[11 rows x 6 columns]

#### **Create new dataframe**

In [28]:
new_dataframe = df.loc[:,'Region':'Net migration']

In [ ]:
#new_dataframe

In [29]:
new_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 6 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   Region                        235 non-null    object
 1   Population                    235 non-null    int64 
 2   Area (sq. mi.)                235 non-null    int64 
 3   Pop. Density (per sq. mi.)    235 non-null    object
 4   Coastline (coast/area ratio)  235 non-null    object
 5   Net migration                 232 non-null    object
dtypes: int64(2), object(4)
memory usage: 11.1+ KB


#### **Locate Rows, Conditional**

**`df.loc[]`**

Access a group of rows and columns by label(s) or a boolean array.

`df.loc[]` is primarily label based, but may also be used with a boolean array.  

Allowed inputs are:

- A single label, e.g. 5 or 'a', (note that 5 is interpreted as a label of the index, and never as an integer position along the index). 

- A list or array of labels, e.g. ['a', 'b', 'c'].  

- A slice object with labels, e.g. 'a':'f'.  

[For more information about df.loc[] click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html)

#### **Locate numeric**

In [30]:
df.loc[df['Area (sq. mi.)'] == 7686850]

Country                               Region  Population  \
12  Australia   OCEANIA                                20264082   

    Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
12         7686850                        2,6  ...     12,14      7,51   

   Agriculture  Industry Service  
12       0,038     0,262     0,7  

[1 rows x 20 columns]

In [31]:
df.loc[df['Area (sq. mi.)']>7000000]

Country                               Region  Population  \
12       Australia   OCEANIA                                20264082   
28          Brazil               LATIN AMER. & CARIB       188078227   
29          Brazil               LATIN AMER. & CARIB       188078227   
38          Canada   NORTHERN AMERICA                       33098932   
44           China         ASIA (EX. NEAR EAST)           1313973713   
174         Russia                  C.W. OF IND. STATES    142893540   
222  United States   NORTHERN AMERICA                      298444215   

     Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
12          7686850                        2,6  ...     12,14      7,51   
28          8511965                       22,1  ...     16,56      6,17   
29          8511965                       22,1  ...     16,56      6,17   
38          9984670                        3,3  ...     10,78       7,8   
44          9596960                      136,9  ...     13,25      6,97   
174        17075200                        8,4  ...      9,95     14,65   
222         9631420                       31,0  ...     14,14      8,26   

    Agriculture  Industry Service  
12        0,038     0,262     0,7  
28        0,084       0,4   0,516  
29        0,084       0,4   0,516  
38        0,022     0,294   0,684  
44        0,125     0,473   0,403  
174       0,054     0,371   0,575  
222        0,01     0,204   0,787  

[7 rows x 20 columns]

#### **Locate string**

In [ ]:
df.loc[df['Region'].str.contains("OCEANIA")]

Country                               Region  Population  \
3       American Samoa   OCEANIA                                   57794   
12           Australia   OCEANIA                                20264082   
49        Cook Islands   OCEANIA                                   21388   
70                Fiji   OCEANIA                                  905949   
74    French Polynesia   OCEANIA                                  274578   
..                  ...                                  ...         ...   
211              Tonga   OCEANIA                                  114689   
212              Tonga   OCEANIA                                  114689   
218             Tuvalu   OCEANIA                                   11810   
225            Vanuatu   OCEANIA                                  208869   
229  Wallis and Futuna   OCEANIA                                   16025   

     Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
3               199                      290,4  ...     22,46      3,27   
12          7686850                        2,6  ...     12,14      7,51   
49              240                       89,1  ...        21       NaN   
70            18270                       49,6  ...     22,55      5,65   
74             4167                       65,9  ...     16,68      4,69   
..              ...                        ...  ...       ...       ...   
211             748                      153,3  ...     25,37      5,28   
212             748                      153,3  ...     25,37      5,28   
218              26                      454,2  ...     22,18      7,11   
225           12200                       17,1  ...     22,72      7,82   
229             274                       58,5  ...       NaN       NaN   

    Agriculture  Industry Service  
3           NaN       NaN     NaN  
12        0,038     0,262     0,7  
49        0,151     0,096   0,753  
70        0,089     0,135   0,776  
74        0,031      0,19   0,769  
..          ...       ...     ...  
211        0,23      0,27     0,5  
212        0,23      0,27     0,5  
218       0,166     0,272   0,562  
225        0,26      0,12    0,62  
229         NaN       NaN     NaN  

[22 rows x 20 columns]

In [ ]:
df.loc[df['Region'].str.contains("oceania", case = False)]

Country                               Region  Population  \
3       American Samoa   OCEANIA                                   57794   
12           Australia   OCEANIA                                20264082   
49        Cook Islands   OCEANIA                                   21388   
70                Fiji   OCEANIA                                  905949   
74    French Polynesia   OCEANIA                                  274578   
..                  ...                                  ...         ...   
211              Tonga   OCEANIA                                  114689   
212              Tonga   OCEANIA                                  114689   
218             Tuvalu   OCEANIA                                   11810   
225            Vanuatu   OCEANIA                                  208869   
229  Wallis and Futuna   OCEANIA                                   16025   

     Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
3               199                      290,4  ...     22,46      3,27   
12          7686850                        2,6  ...     12,14      7,51   
49              240                       89,1  ...        21       NaN   
70            18270                       49,6  ...     22,55      5,65   
74             4167                       65,9  ...     16,68      4,69   
..              ...                        ...  ...       ...       ...   
211             748                      153,3  ...     25,37      5,28   
212             748                      153,3  ...     25,37      5,28   
218              26                      454,2  ...     22,18      7,11   
225           12200                       17,1  ...     22,72      7,82   
229             274                       58,5  ...       NaN       NaN   

    Agriculture  Industry Service  
3           NaN       NaN     NaN  
12        0,038     0,262     0,7  
49        0,151     0,096   0,753  
70        0,089     0,135   0,776  
74        0,031      0,19   0,769  
..          ...       ...     ...  
211        0,23      0,27     0,5  
212        0,23      0,27     0,5  
218       0,166     0,272   0,562  
225        0,26      0,12    0,62  
229         NaN       NaN     NaN  

[22 rows x 20 columns]

[For more information about string.contains() click here](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html)

#### **Locating by index**

#### **df.iloc**
property DataFrame.iloc[source]

Purely integer-location based indexing for selection by position.

.iloc[] is primarily integer position based (from 0 to length-1 of the axis), but may also be used with a boolean array.

#### df.iloc Syntax
##### `df.iloc[index_of_rows, index_of_columns]`

In [34]:
df.iloc[0:10]

Country                               Region  Population  \
0        Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1            Albania   EASTERN EUROPE                          3581655   
2            Algeria   NORTHERN AFRICA                        32930091   
3     American Samoa   OCEANIA                                   57794   
4            Andorra   WESTERN EUROPE                            71201   
5             Angola   SUB-SAHARAN AFRICA                     12127071   
6           Anguilla               LATIN AMER. & CARIB           13477   
7  Antigua & Barbuda               LATIN AMER. & CARIB           69108   
8  Antigua & Barbuda               LATIN AMER. & CARIB           69108   
9          Argentina               LATIN AMER. & CARIB        39921833   

   Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
0          647500                       48,0  ...      46,6     20,34   
1           28748                      124,6  ...     15,11      5,22   
2         2381740                       13,8  ...     17,14      4,61   
3             199                      290,4  ...     22,46      3,27   
4             468                      152,1  ...      8,71      6,25   
5         1246700                        9,7  ...     45,11      24,2   
6             102                      132,1  ...     14,17      5,34   
7             443                      156,0  ...     16,93      5,37   
8             443                      156,0  ...     16,93      5,37   
9         2766890                       14,4  ...     16,73      7,55   

  Agriculture  Industry Service  
0        0,38      0,24    0,38  
1       0,232     0,188   0,579  
2       0,101       0,6   0,298  
3         NaN       NaN     NaN  
4         NaN       NaN     NaN  
5       0,096     0,658   0,246  
6        0,04      0,18    0,78  
7       0,038      0,22   0,743  
8       0,038      0,22   0,743  
9       0,095     0,358   0,547  

[10 rows x 20 columns]

In [35]:
df.iloc[0:5,1:4] 

Region  Population  Area (sq. mi.)
0        ASIA (EX. NEAR EAST)             31056997          647500
1  EASTERN EUROPE                          3581655           28748
2  NORTHERN AFRICA                        32930091         2381740
3  OCEANIA                                   57794             199
4  WESTERN EUROPE                            71201             468

# Drop rows

`df=df.drop(row_index,axis=0)`

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 20 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Country                             235 non-null    object 
 1   Region                              235 non-null    object 
 2   Population                          235 non-null    int64  
 3   Area (sq. mi.)                      235 non-null    int64  
 4   Pop. Density (per sq. mi.)          235 non-null    object 
 5   Coastline (coast/area ratio)        235 non-null    object 
 6   Net migration                       232 non-null    object 
 7   Infant mortality (per 1000 births)  232 non-null    object 
 8   GDP ($ per capita)                  234 non-null    float64
 9   Literacy (%)                        216 non-null    object 
 10  Phones (per 1000)                   230 non-null    object 
 11  Arable (%)                          232 non-n

[For more information about df.info() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html)

# **Working with nulls**

#### **Identifying Nulls**

We can clearly see that though there are 235 rows in the table not all columns contain 235 non-null objects. This fact implies that there are Nulls present in those columns.  
Let's look for nulls with df.isnull()  
[For more information about df.isnull() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isnull.html)


In [37]:
df.isnull() # This is not, in itself, super useful

Country  Region  Population  Area (sq. mi.)  Pop. Density (per sq. mi.)  \
0      False   False       False           False                       False   
1      False   False       False           False                       False   
2      False   False       False           False                       False   
3      False   False       False           False                       False   
4      False   False       False           False                       False   
..       ...     ...         ...             ...                         ...   
230    False   False       False           False                       False   
231    False   False       False           False                       False   
232    False   False       False           False                       False   
233    False   False       False           False                       False   
234    False   False       False           False                       False   

     ...  Birthrate  Deathrate  Agriculture  Industry  Service  
0    ...      False      False        False     False    False  
1    ...      False      False        False     False    False  
2    ...      False      False        False     False    False  
3    ...      False      False         True      True     True  
4    ...      False      False         True      True     True  
..   ...        ...        ...          ...       ...      ...  
230  ...      False      False        False     False    False  
231  ...       True       True         True      True    False  
232  ...      False      False        False     False    False  
233  ...      False      False        False     False    False  
234  ...      False      False        False     False    False  

[235 rows x 20 columns]

In [38]:
df.isnull().sum() # .sum() works because True is coded as 1 and False is coded as 0

Country                        0
Region                         0
Population                     0
Area (sq. mi.)                 0
Pop. Density (per sq. mi.)     0
                              ..
Birthrate                      3
Deathrate                      4
Agriculture                   16
Industry                      17
Service                       16
Length: 20, dtype: int64

In [39]:
df.isnull().sum().sum() # gives us a total aggregate of nulls

118

#### **Display rows with nulls**

##### For entire dataframe

In [42]:
#df[df.isnull().any(axis=1)]
#df[df.isna().any(axis=1)]

Country                               Region  Population  \
3          American Samoa   OCEANIA                                   57794   
4                 Andorra   WESTERN EUROPE                            71201   
5                  Angola   SUB-SAHARAN AFRICA                     12127071   
26   Bosnia & Herzegovina   EASTERN EUROPE                          4498976   
38                 Canada   NORTHERN AMERICA                       33098932   
..                     ...                                  ...         ...   
218                Tuvalu   OCEANIA                                   11810   
228        Virgin Islands               LATIN AMER. & CARIB          108605   
229     Wallis and Futuna   OCEANIA                                   16025   
230             West Bank   NEAR EAST                               2460492   
231        Western Sahara   NORTHERN AFRICA                          273008   

     Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
3               199                      290,4  ...     22,46      3,27   
4               468                      152,1  ...      8,71      6,25   
5           1246700                        9,7  ...     45,11      24,2   
26            51129                       88,0  ...      8,77      8,27   
38          9984670                        3,3  ...     10,78       7,8   
..              ...                        ...  ...       ...       ...   
218              26                      454,2  ...     22,18      7,11   
228            1910                       56,9  ...     13,96      6,43   
229             274                       58,5  ...       NaN       NaN   
230            5860                      419,9  ...     31,67      3,92   
231          266000                        1,0  ...       NaN       NaN   

    Agriculture  Industry Service  
3           NaN       NaN     NaN  
4           NaN       NaN     NaN  
5         0,096     0,658   0,246  
26        0,142     0,308    0,55  
38        0,022     0,294   0,684  
..          ...       ...     ...  
218       0,166     0,272   0,562  
228        0,01      0,19     0,8  
229         NaN       NaN     NaN  
230        0,09      0,28    0,63  
231         NaN       NaN     0,4  

[50 rows x 20 columns]

##### For a column

In [47]:
df[df['Deathrate'].isnull()]
#df[df['Deathrate'].isna()]

Country                               Region  Population  \
49        Cook Islands   OCEANIA                                   21388   
187             Serbia   EASTERN EUROPE                          9396411   
229  Wallis and Futuna   OCEANIA                                   16025   
231     Western Sahara   NORTHERN AFRICA                          273008   

     Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
49              240                       89,1  ...        21       NaN   
187           88361                      106,3  ...       NaN       NaN   
229             274                       58,5  ...       NaN       NaN   
231          266000                        1,0  ...       NaN       NaN   

    Agriculture  Industry Service  
49        0,151     0,096   0,753  
187       0,166     0,255   0,579  
229         NaN       NaN     NaN  
231         NaN       NaN     0,4  

[4 rows x 20 columns]

#### **Replacing Nulls**

`df.fillna()` replaces NA/NaN values using the specified method.

[For more information about df.fillna() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html)

In [52]:
df1 = df.fillna(value = '0')

In [53]:
df1

Country                               Region  Population  \
0       Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1           Albania   EASTERN EUROPE                          3581655   
2           Algeria   NORTHERN AFRICA                        32930091   
3    American Samoa   OCEANIA                                   57794   
4           Andorra   WESTERN EUROPE                            71201   
..               ...                                  ...         ...   
230       West Bank   NEAR EAST                               2460492   
231  Western Sahara   NORTHERN AFRICA                          273008   
232           Yemen   NEAR EAST                              21456188   
233          Zambia   SUB-SAHARAN AFRICA                     11502010   
234        Zimbabwe   SUB-SAHARAN AFRICA                     12236805   

     Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
0            647500                       48,0  ...      46,6     20,34   
1             28748                      124,6  ...     15,11      5,22   
2           2381740                       13,8  ...     17,14      4,61   
3               199                      290,4  ...     22,46      3,27   
4               468                      152,1  ...      8,71      6,25   
..              ...                        ...  ...       ...       ...   
230            5860                      419,9  ...     31,67      3,92   
231          266000                        1,0  ...         0         0   
232          527970                       40,6  ...     42,89       8,3   
233          752614                       15,3  ...        41     19,93   
234          390580                       31,3  ...     28,01     21,84   

    Agriculture Industry Service  
0          0,38     0,24    0,38  
1         0,232    0,188   0,579  
2         0,101      0,6   0,298  
3             0        0       0  
4             0        0       0  
..          ...      ...     ...  
230        0,09     0,28    0,63  
231           0        0     0,4  
232       0,135    0,472   0,393  
233        0,22     0,29   0,489  
234       0,179    0,243   0,579  

[235 rows x 20 columns]

In [54]:
df1.isnull().sum().sum()

0

In [55]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 20 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Country                             235 non-null    object
 1   Region                              235 non-null    object
 2   Population                          235 non-null    int64 
 3   Area (sq. mi.)                      235 non-null    int64 
 4   Pop. Density (per sq. mi.)          235 non-null    object
 5   Coastline (coast/area ratio)        235 non-null    object
 6   Net migration                       235 non-null    object
 7   Infant mortality (per 1000 births)  235 non-null    object
 8   GDP ($ per capita)                  235 non-null    object
 9   Literacy (%)                        235 non-null    object
 10  Phones (per 1000)                   235 non-null    object
 11  Arable (%)                          235 non-null    object

#### **Removing Nulls** 

`df.dropna()` removes columns or rows containing missing values.

In [56]:
df2 = df.dropna()

In [57]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 185 entries, 0 to 234
Data columns (total 20 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Country                             185 non-null    object 
 1   Region                              185 non-null    object 
 2   Population                          185 non-null    int64  
 3   Area (sq. mi.)                      185 non-null    int64  
 4   Pop. Density (per sq. mi.)          185 non-null    object 
 5   Coastline (coast/area ratio)        185 non-null    object 
 6   Net migration                       185 non-null    object 
 7   Infant mortality (per 1000 births)  185 non-null    object 
 8   GDP ($ per capita)                  185 non-null    float64
 9   Literacy (%)                        185 non-null    object 
 10  Phones (per 1000)                   185 non-null    object 
 11  Arable (%)                          185 non-n

[For more information on the df.dropna() method click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html)

#### **Identifying Duplicates**

`df.duplicated()` returns boolean Series denoting duplicate rows.

[For more information about df.duplicated() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html)

In [58]:
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
230    False
231    False
232    False
233    False
234    False
Length: 235, dtype: bool

In [59]:
df.duplicated().sum()

9

#### **Extract duplicates**

In [61]:
#df.loc[df.duplicated()]
#df.loc[df.duplicated(),'Population':'Net migration']

Population  Area (sq. mi.) Pop. Density (per sq. mi.)  \
8         69108             443                      156,0   
29    188078227         8511965                       22,1   
62     13547510          283560                       47,8   
89        65409              78                      838,6   
145        9439             102                       92,5   
176     8648248           26338                      328,4   
200      439117          163270                        2,7   
205    18881361          185180                      102,0   
212      114689             748                      153,3   

    Coastline (coast/area ratio) Net migration  
8                          34,54         -6,15  
29                          0,09         -0,03  
62                          0,79         -8,58  
89                         64,10          3,84  
145                        39,22             0  
176                         0,00             0  
200                         0,24         -8,81  
205                         0,10             0  
212                        56,02             0

#### **Removing Duplicates**

`df.drop_duplicates()` returns DataFrame with duplicate rows removed.

In [68]:
df_no_dupes = df.drop_duplicates()

In [69]:
df_no_dupes.duplicated().sum()

0

In [71]:
df_no_dupes.shape

(226, 20)

In [72]:
df_no_dupes

Country                               Region  Population  \
0       Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1           Albania   EASTERN EUROPE                          3581655   
2           Algeria   NORTHERN AFRICA                        32930091   
3    American Samoa   OCEANIA                                   57794   
4           Andorra   WESTERN EUROPE                            71201   
..               ...                                  ...         ...   
230       West Bank   NEAR EAST                               2460492   
231  Western Sahara   NORTHERN AFRICA                          273008   
232           Yemen   NEAR EAST                              21456188   
233          Zambia   SUB-SAHARAN AFRICA                     11502010   
234        Zimbabwe   SUB-SAHARAN AFRICA                     12236805   

     Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
0            647500                       48,0  ...      46,6     20,34   
1             28748                      124,6  ...     15,11      5,22   
2           2381740                       13,8  ...     17,14      4,61   
3               199                      290,4  ...     22,46      3,27   
4               468                      152,1  ...      8,71      6,25   
..              ...                        ...  ...       ...       ...   
230            5860                      419,9  ...     31,67      3,92   
231          266000                        1,0  ...       NaN       NaN   
232          527970                       40,6  ...     42,89       8,3   
233          752614                       15,3  ...        41     19,93   
234          390580                       31,3  ...     28,01     21,84   

    Agriculture  Industry Service  
0          0,38      0,24    0,38  
1         0,232     0,188   0,579  
2         0,101       0,6   0,298  
3           NaN       NaN     NaN  
4           NaN       NaN     NaN  
..          ...       ...     ...  
230        0,09      0,28    0,63  
231         NaN       NaN     0,4  
232       0,135     0,472   0,393  
233        0,22      0,29   0,489  
234       0,179     0,243   0,579  

[226 rows x 20 columns]

In [73]:
df_no_dupes.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 226 entries, 0 to 234
Data columns (total 20 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Country                             226 non-null    object 
 1   Region                              226 non-null    object 
 2   Population                          226 non-null    int64  
 3   Area (sq. mi.)                      226 non-null    int64  
 4   Pop. Density (per sq. mi.)          226 non-null    object 
 5   Coastline (coast/area ratio)        226 non-null    object 
 6   Net migration                       223 non-null    object 
 7   Infant mortality (per 1000 births)  223 non-null    object 
 8   GDP ($ per capita)                  225 non-null    float64
 9   Literacy (%)                        208 non-null    object 
 10  Phones (per 1000)                   222 non-null    object 
 11  Arable (%)                          224 non-n

In [74]:
df_no_dupes

Country                               Region  Population  \
0       Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1           Albania   EASTERN EUROPE                          3581655   
2           Algeria   NORTHERN AFRICA                        32930091   
3    American Samoa   OCEANIA                                   57794   
4           Andorra   WESTERN EUROPE                            71201   
..               ...                                  ...         ...   
230       West Bank   NEAR EAST                               2460492   
231  Western Sahara   NORTHERN AFRICA                          273008   
232           Yemen   NEAR EAST                              21456188   
233          Zambia   SUB-SAHARAN AFRICA                     11502010   
234        Zimbabwe   SUB-SAHARAN AFRICA                     12236805   

     Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
0            647500                       48,0  ...      46,6     20,34   
1             28748                      124,6  ...     15,11      5,22   
2           2381740                       13,8  ...     17,14      4,61   
3               199                      290,4  ...     22,46      3,27   
4               468                      152,1  ...      8,71      6,25   
..              ...                        ...  ...       ...       ...   
230            5860                      419,9  ...     31,67      3,92   
231          266000                        1,0  ...       NaN       NaN   
232          527970                       40,6  ...     42,89       8,3   
233          752614                       15,3  ...        41     19,93   
234          390580                       31,3  ...     28,01     21,84   

    Agriculture  Industry Service  
0          0,38      0,24    0,38  
1         0,232     0,188   0,579  
2         0,101       0,6   0,298  
3           NaN       NaN     NaN  
4           NaN       NaN     NaN  
..          ...       ...     ...  
230        0,09      0,28    0,63  
231         NaN       NaN     0,4  
232       0,135     0,472   0,393  
233        0,22      0,29   0,489  
234       0,179     0,243   0,579  

[226 rows x 20 columns]

In [66]:
df3.head(10)

Country                               Region  Population  \
0         Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1             Albania   EASTERN EUROPE                          3581655   
2             Algeria   NORTHERN AFRICA                        32930091   
3      American Samoa   OCEANIA                                   57794   
4             Andorra   WESTERN EUROPE                            71201   
5              Angola   SUB-SAHARAN AFRICA                     12127071   
6            Anguilla               LATIN AMER. & CARIB           13477   
7   Antigua & Barbuda               LATIN AMER. & CARIB           69108   
9           Argentina               LATIN AMER. & CARIB        39921833   
10            Armenia                  C.W. OF IND. STATES      2976372   

    Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
0           647500                       48,0  ...      46,6     20,34   
1            28748                      124,6  ...     15,11      5,22   
2          2381740                       13,8  ...     17,14      4,61   
3              199                      290,4  ...     22,46      3,27   
4              468                      152,1  ...      8,71      6,25   
5          1246700                        9,7  ...     45,11      24,2   
6              102                      132,1  ...     14,17      5,34   
7              443                      156,0  ...     16,93      5,37   
9          2766890                       14,4  ...     16,73      7,55   
10           29800                       99,9  ...     12,07      8,23   

   Agriculture  Industry Service  
0         0,38      0,24    0,38  
1        0,232     0,188   0,579  
2        0,101       0,6   0,298  
3          NaN       NaN     NaN  
4          NaN       NaN     NaN  
5        0,096     0,658   0,246  
6         0,04      0,18    0,78  
7        0,038      0,22   0,743  
9        0,095     0,358   0,547  
10       0,239     0,343   0,418  

[10 rows x 20 columns]

##### Removing the duplicate rows resulted in a non sequential index where index 8 is missing. To reset the indes use `df.reset_index()`

In [76]:
df_no_dupes.reset_index(inplace = True)
df_no_dupes

index          Country                               Region  Population  \
0        0     Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1        1         Albania   EASTERN EUROPE                          3581655   
2        2         Algeria   NORTHERN AFRICA                        32930091   
3        3  American Samoa   OCEANIA                                   57794   
4        4         Andorra   WESTERN EUROPE                            71201   
..     ...              ...                                  ...         ...   
221    230       West Bank   NEAR EAST                               2460492   
222    231  Western Sahara   NORTHERN AFRICA                          273008   
223    232           Yemen   NEAR EAST                              21456188   
224    233          Zambia   SUB-SAHARAN AFRICA                     11502010   
225    234        Zimbabwe   SUB-SAHARAN AFRICA                     12236805   

     Area (sq. mi.)  ... Birthrate Deathrate Agriculture Industry  Service  
0            647500  ...      46,6     20,34        0,38     0,24     0,38  
1             28748  ...     15,11      5,22       0,232    0,188    0,579  
2           2381740  ...     17,14      4,61       0,101      0,6    0,298  
3               199  ...     22,46      3,27         NaN      NaN      NaN  
4               468  ...      8,71      6,25         NaN      NaN      NaN  
..              ...  ...       ...       ...         ...      ...      ...  
221            5860  ...     31,67      3,92        0,09     0,28     0,63  
222          266000  ...       NaN       NaN         NaN      NaN      0,4  
223          527970  ...     42,89       8,3       0,135    0,472    0,393  
224          752614  ...        41     19,93        0,22     0,29    0,489  
225          390580  ...     28,01     21,84       0,179    0,243    0,579  

[226 rows x 21 columns]

# **Transforming the data**

#### **Removing columns**

In [81]:
#df_no_dupes = df_no_dupes.drop(columns = ['level_0','index']) 
df_no_dupes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226 entries, 0 to 225
Data columns (total 21 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   index                               226 non-null    int64  
 1   Country                             226 non-null    object 
 2   Region                              226 non-null    object 
 3   Population                          226 non-null    int64  
 4   Area (sq. mi.)                      226 non-null    int64  
 5   Pop. Density (per sq. mi.)          226 non-null    object 
 6   Coastline (coast/area ratio)        226 non-null    object 
 7   Net migration                       223 non-null    object 
 8   Infant mortality (per 1000 births)  223 non-null    object 
 9   GDP ($ per capita)                  225 non-null    float64
 10  Literacy (%)                        208 non-null    object 
 11  Phones (per 1000)                   222 non-n

Create a new dataframe 'df1 by using the **df.drop()** function to remove the following columns: 'Arable (%)', 'Crops (%)', 'Other (%)', 'Climate', 'Agriculture'

`df1 = df.drop(columns = ['Column Name 1', 'Column Name 2', ...])`

[For more information about df.drop() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop.html)

In [83]:
#df1 = df.drop(columns =['Arable (%)', 'Crops (%)', 'Other (%)', 'Climate', 'Agriculture'])
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 15 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Country                             235 non-null    object 
 1   Region                              235 non-null    object 
 2   Population                          235 non-null    int64  
 3   Area (sq. mi.)                      235 non-null    int64  
 4   Pop. Density (per sq. mi.)          235 non-null    object 
 5   Coastline (coast/area ratio)        235 non-null    object 
 6   Net migration                       232 non-null    object 
 7   Infant mortality (per 1000 births)  232 non-null    object 
 8   GDP ($ per capita)                  234 non-null    float64
 9   Literacy (%)                        216 non-null    object 
 10  Phones (per 1000)                   230 non-null    object 
 11  Birthrate                           232 non-n

#### **Converting data to appropriate type**

Replace comma (,) character with the period (.) for all cells in the ' Pop. Density (per sq. mi.)' column
- df['column name'] = df['column name'].str.replace('old character','new character')  
- Use **df.head()** to confirm results



In [84]:
df1['Pop. Density (per sq. mi.)'] = df1['Pop. Density (per sq. mi.)'].str.replace(',','.')

In [85]:
df1.head()

Country                               Region  Population  \
0     Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1         Albania   EASTERN EUROPE                          3581655   
2         Algeria   NORTHERN AFRICA                        32930091   
3  American Samoa   OCEANIA                                   57794   
4         Andorra   WESTERN EUROPE                            71201   

   Area (sq. mi.) Pop. Density (per sq. mi.)  ... Phones (per 1000) Birthrate  \
0          647500                       48.0  ...               3,2      46,6   
1           28748                      124.6  ...              71,2     15,11   
2         2381740                       13.8  ...              78,1     17,14   
3             199                      290.4  ...             259,5     22,46   
4             468                      152.1  ...             497,2      8,71   

  Deathrate  Industry Service  
0     20,34      0,24    0,38  
1      5,22     0,188   0,579  
2      4,61       0,6   0,298  
3      3,27       NaN     NaN  
4      6,25       NaN     NaN  

[5 rows x 15 columns]

Because the comma (,) is used exclusively in numeric fields we can replace it for the whole table

- df = df.replace('old character','new character', regex=True)
- Use df.head() to confirm results

In [87]:
df1 = df1.replace(',','.', regex=True)
df1.head(2)

Country                               Region  Population  \
0  Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1      Albania   EASTERN EUROPE                          3581655   

   Area (sq. mi.) Pop. Density (per sq. mi.)  ... Phones (per 1000) Birthrate  \
0          647500                       48.0  ...               3.2      46.6   
1           28748                      124.6  ...              71.2     15.11   

  Deathrate  Industry Service  
0     20.34      0.24    0.38  
1      5.22     0.188   0.579  

[2 rows x 15 columns]

#### **Converting Type for single column**

In [88]:
# convert "Pop Density" from string to float
df1['Pop. Density (per sq. mi.)'] = df1['Pop. Density (per sq. mi.)'].astype(float)

In [89]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 15 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Country                             235 non-null    object 
 1   Region                              235 non-null    object 
 2   Population                          235 non-null    int64  
 3   Area (sq. mi.)                      235 non-null    int64  
 4   Pop. Density (per sq. mi.)          235 non-null    float64
 5   Coastline (coast/area ratio)        235 non-null    object 
 6   Net migration                       232 non-null    object 
 7   Infant mortality (per 1000 births)  232 non-null    object 
 8   GDP ($ per capita)                  234 non-null    float64
 9   Literacy (%)                        216 non-null    object 
 10  Phones (per 1000)                   230 non-null    object 
 11  Birthrate                           232 non-n

#### **Convert multiple columns** 

#### Would change whole table * if compatible with datatype
DataFrame.astype(dtype)

In [90]:
df1 = df1.astype({'Coastline (coast/area ratio)':'float','Net migration':'float','Infant mortality (per 1000 births)':'float','GDP ($ per capita)':'float','Literacy (%)':'float','Phones (per 1000)':'float','Birthrate':'float','Deathrate':'float'})

In [92]:
df.columns

Index(['Country', 'Region', 'Population', 'Area (sq. mi.)',
       'Pop. Density (per sq. mi.)', 'Coastline (coast/area ratio)',
       'Net migration', 'Infant mortality (per 1000 births)',
       'GDP ($ per capita)', 'Literacy (%)', 'Phones (per 1000)', 'Arable (%)',
       'Crops (%)', 'Other (%)', 'Climate', 'Birthrate', 'Deathrate',
       'Agriculture', 'Industry', 'Service'],
      dtype='object')

In [91]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 15 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Country                             235 non-null    object 
 1   Region                              235 non-null    object 
 2   Population                          235 non-null    int64  
 3   Area (sq. mi.)                      235 non-null    int64  
 4   Pop. Density (per sq. mi.)          235 non-null    float64
 5   Coastline (coast/area ratio)        235 non-null    float64
 6   Net migration                       232 non-null    float64
 7   Infant mortality (per 1000 births)  232 non-null    float64
 8   GDP ($ per capita)                  234 non-null    float64
 9   Literacy (%)                        216 non-null    float64
 10  Phones (per 1000)                   230 non-null    float64
 11  Birthrate                           232 non-n

In [93]:
df1.describe()

Population  Area (sq. mi.)  Pop. Density (per sq. mi.)  \
count  2.350000e+02    2.350000e+02                  235.000000   
mean   2.872891e+07    6.165375e+05                  373.428511   
std    1.164019e+08    1.836074e+06                 1632.507460   
min    7.026000e+03    2.000000e+00                    0.000000   
25%    4.105965e+05    3.488500e+03                   29.150000   
50%    4.786994e+06    8.660000e+04                   79.800000   
75%    1.749777e+07    4.219110e+05                  190.150000   
max    1.313974e+09    1.707520e+07                16271.500000   

       Coastline (coast/area ratio)  Net migration  ...  GDP ($ per capita)  \
count                    235.000000     232.000000  ...          234.000000   
mean                      21.268255      -0.052672  ...         9499.145299   
std                       71.201669       4.894217  ...         9918.790385   
min                        0.000000     -20.990000  ...          500.000000   
25%                        0.100000      -0.997500  ...         1900.000000   
50%                        0.730000       0.000000  ...         5400.000000   
75%                       11.470000       0.955000  ...        14850.000000   
max                      870.660000      23.060000  ...        55100.000000   

       Literacy (%)  Phones (per 1000)   Birthrate   Deathrate  
count    216.000000         230.000000  232.000000  231.000000  
mean      83.050926         236.297826   22.105345    9.189394  
std       19.499965         229.332027   11.106471    4.953983  
min       17.600000           0.200000    7.290000    2.290000  
25%       73.300000          38.450000   12.732500    5.830000  
50%       92.500000         172.900000   18.750000    7.820000  
75%       98.000000         383.900000   29.755000   10.580000  
max      100.000000        1035.600000   50.730000   29.740000  

[8 rows x 11 columns]

In [94]:
# Rounding is performed last on the whole output of the aggregation
print(round(df1['Population'].mean(),2))
print('The mean Area is: ' + str(round(df['Area (sq. mi.)'].mean(),2)))

28728914.66
The mean Area is: 616537.54


## **String Methods**

In [95]:
df1.head(2)

Country                               Region  Population  \
0  Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1      Albania   EASTERN EUROPE                          3581655   

   Area (sq. mi.)  Pop. Density (per sq. mi.)  ...  Phones (per 1000)  \
0          647500                        48.0  ...                3.2   
1           28748                       124.6  ...               71.2   

   Birthrate  Deathrate  Industry  Service  
0      46.60      20.34      0.24     0.38  
1      15.11       5.22     0.188    0.579  

[2 rows x 15 columns]

In [96]:
df1['Region'] = df1['Region'].str.lower()

In [97]:
df1.head()

Country                               Region  Population  \
0     Afghanistan         asia (ex. near east)             31056997   
1         Albania   eastern europe                          3581655   
2         Algeria   northern africa                        32930091   
3  American Samoa   oceania                                   57794   
4         Andorra   western europe                            71201   

   Area (sq. mi.)  Pop. Density (per sq. mi.)  ...  Phones (per 1000)  \
0          647500                        48.0  ...                3.2   
1           28748                       124.6  ...               71.2   
2         2381740                        13.8  ...               78.1   
3             199                       290.4  ...              259.5   
4             468                       152.1  ...              497.2   

   Birthrate  Deathrate  Industry  Service  
0      46.60      20.34      0.24     0.38  
1      15.11       5.22     0.188    0.579  
2      17.14       4.61       0.6    0.298  
3      22.46       3.27       NaN      NaN  
4       8.71       6.25       NaN      NaN  

[5 rows x 15 columns]

**Replace each string in pandas.Series**  
- str.replace()  

**Strip each string in pandas.Series**   
- str.strip()  
- str.lstrip()  
- str.rstrip()  

**Convert the case of each string in pandas.Series** 
- str.lower()  
- str.upper()  
- str.capitalize()  
- str.title()

**Split strings around given separator/delimiter from the beginning, at the specified delimiter string**
- str.split()

**Check, True or False, if record contains string**

- str.contains()


#### How many countries have the name Africa or America in title

In [107]:
df['Country'].str.contains('Africa|America').value_counts()[True]

3

In [108]:
df['Country'].str.contains('Africa|America').sum()

3

#### Show all rows containing string

In [109]:
df[df['Country'].str.contains('Africa|America')]

Country                               Region  Population  \
3          American Samoa   OCEANIA                                   57794   
41   Central African Rep.   SUB-SAHARAN AFRICA                      4303356   
195          South Africa   SUB-SAHARAN AFRICA                     44187637   

     Area (sq. mi.) Pop. Density (per sq. mi.)  ... Birthrate Deathrate  \
3               199                      290,4  ...     22,46      3,27   
41           622984                        6,9  ...     33,91     18,65   
195         1219912                       36,2  ...      18,2        22   

    Agriculture  Industry Service  
3           NaN       NaN     NaN  
41         0,55       0,2    0,25  
195       0,025     0,303   0,671  

[3 rows x 20 columns]

## **Joins**

Join columns with other DataFrame either on index or on a key column. Efficiently join multiple DataFrame objects by index at once by passing a list. Left join by default.

* Uses the first row of the right table as index

[For more information on joins click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.join.html)

In [ ]:
ss_orders = pd.read_csv('data/sample_superstore/sample_superstore_orders.csv')

In [ ]:
ss_returns = pd.read_csv('data/sample_superstore/sample_superstore_returns.csv')

In [ ]:
joined_df = ss_orders.join(ss_returns.set_index('Order ID'), on='Order ID', how = 'inner')

**How though?**

`how` accepts arguments that define how the tables are joined.

- left: use calling frame’s index (or column if on is specified)
- right: use other’s index.
- outer: form union of calling frame’s index (or column if on is specified) with other’s index, and sort it. lexicographically.
- inner: form intersection of calling frame’s index (or column if on is specified) with other’s index, preserving the order of the calling’s one.
- cross: creates the cartesian product from both frames, preserves the order of the left keys.

## **Merge**

- For combining data on common columns
- More flexible, and more complex than the Join method
- Many to one joins possible
- Side by side merge
- Inner join by default

In [ ]:
ss_orders = pd.read_csv('data/sample_superstore/sample_superstore_orders.csv')

In [ ]:
ss_returns = pd.read_csv('data/sample_superstore/sample_superstore_returns.csv')

In [ ]:
ss_merged = pd.merge(ss_orders, ss_returns)
ss_merged.head(1)

In [ ]:
ss_merged_on_order_id = pd.merge(ss_orders, ss_returns, on = 'Order ID')
ss_merged_on_order_id.head(1)

## **Pivot Table**

Produce a “pivot” table based on column values. Uses unique values from specified index / columns to form axes of the resulting DataFrame. This function does not support data aggregation, multiple values will result in a MultiIndex in the columns. 

- `df.pivot()` reshapes data based on clumns 

[For more information about df.pivot() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pivot.html)

- `df.pivot_table()` creates a pivot table as a dataframe allowing aggregation and tabulation

[For more information about df.pivot_table() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pivot_table.html)

In [ ]:
joined_df.head(1)

`df.pivot_table('what to aggregate', ['Column1','Column2'], aggfunc = aggregation)`

In [ ]:
pivot_table = joined_df.pivot_table('Sales', ['Region','State']) #default aggregation is mean
pivot_table.head()

`df.reset_index()` Resets the index, or a level of it.
Reset the index of the DataFrame, and use the default one instead. If the DataFrame has a MultiIndex, this method can remove one or more levels.

In [ ]:
pivot_table.reset_index().head()

[To learn more about df.reset_index() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html)

In [ ]:
pivot_table1 = joined_df.pivot_table('Profit',['Category','Sub-Category'], aggfunc = ['max','min','sum'])
pivot_table1.head()

## **Group By**

A groupby operation involves some combination of splitting the object, applying a function, and combining the results. This can be used to group large amounts of data and compute operations on these groups.

In [ ]:
df1 = joined_df.groupby('Category')[['Profit','Sales']].sum()
df1

In [ ]:
df1.reset_index()

[To learn more about df.groupby() click here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html)

In [ ]:
df2 = joined_df.groupby(['Category','Sub-Category'])[['Profit','Sales']].sum()
df2

#### **Exporting data, the most important step?**

In [ ]:
# df.to_csv('data/new_and_imprved_dataset.csv')

#### **Appendix**

#### **Python Operators**

Before we get into datatypes, we'll talk about operators, which are how you will interact with each data type in python.

Python operators are divided into these groups:

* Arithmetic operators - used with numeric values to perform common mathematical operations
    * `+`, `-`, `*` (multiplication), `**` (exponentiation), `/` (float division), `//` (integer/floor division), `%` (modulo) 
* Assignment operators - used to assign values to variables
    * `=`, `+=` etc.
* Comparison operators - used to compare two variables
    * `==`, (equal) `!=` (not equal), `<=` (less than or equal), `<`, etc.
* Logical operators - used to combine conditional statements
    * `and`, `or`, `not`
* Identity operators - test if two objects are actually the same object (the same location in memory)
    * `is`, `is not`
* Membership operators - test if a sequence is present in an object
    * `in`, `not in`
* Bitwise operators - compare two binary (boolean) numbers
    * `&` (and), `|` (or), `~` (not), and others